### PR TITLE
Expose connection info via API

### PR DESCRIPTION
**Pull request by mikeminutillo**
A companion to https://github.com/Particular/ServiceControl/pull/2669

This creates a new api at `/api/connection` which exposes the info needed for the platform connection package.

```json
{
  "Heartbeats": {
    "Enabled": true,
    "HeartbeatsQueue": "Particular.ServiceControl",
    "Frequency": "00:00:10",
    "TimeToLive": "00:00:40"
  },
  "CustomChecks": {
    "Enabled": true,
    "CustomChecksQueue": "Particular.ServiceControl"
  },
  "ErrorQueue": "error",
  "MessageAudit": {
    "Enabled": true,
    "AuditQueue": "audit"
  },
  "SagaAudit": {
    "Enabled": true,
    "SagaAuditQueue": "audit"
  }
}
```

NOTE: The `MessageAudit` and `SagaAudit` entries come from the audit instance. If you have multiple audit instances then subsequent ones will be rendered as `MessageAudit1`, `MessageAudit2`, and so on. The order that this happens is not defined.

### CODE REVIEW

**Comment by mikeminutillo** [comment]
The acceptance tests are failing because we change the folder structure on the build server. The compile step finds all of the AcceptanceTest assemblies and exports them as artifacts. When the tests run in a later step, they don't have access to the ApprovalFiles. Not only that, but because we move the assemblies to a new location, the approver tries to write to a folder that it doesn't have permission to.

**Comment by mikeminutillo** [comment]
We resolved the issue with the approval tests inside of an acceptance testing project on the build server by:
- Moving the acceptance test DLL to a location at least 3 deep within the build artifacts (Particular.Approvals is assuming this)
- Copying the ApprovalFiles folder into the correct location within the build artifacts

We also resolved some of the issues with the approval tests that work locally on a developer machine that do not work the same way on the build server. i.e. SQL database name is different, SQS Prefix

**Comment by mikeminutillo** [comment]
The SQS test is still broken because the local dev environment connection string includes a queue prefix and the build server connection string does not. We will need to update the build server connection string to include a queue prefix.

File: src/ServiceControl.Audit/Connection/ConnectionController.cs
Comment (tmasternak) [inline src/ServiceControl.Audit/Connection/ConnectionController.cs:17]: Should we have a separate type for the result and approvals to ensure we detect any breaking changes?

File: src/ServiceControl.MultiInstance.AcceptanceTests/Connection/PlatformConnectionTests.cs
Comment (tmasternak) [inline src/ServiceControl.MultiInstance.AcceptanceTests/Connection/PlatformConnectionTests.cs:88]: Should we detect LearningTransport explicitly and exit `Scrub` early in such a case?

File: src/ServiceControl/Connection/PlatformConnectionBuilder.cs
Comment (tmasternak) [inline src/ServiceControl/Connection/PlatformConnectionBuilder.cs:17]: This is [done already](https://github.com/Particular/ServiceControl/pull/2674/files#diff-8b54e2ea88c21bb9db165c40c2a65eb0f58d8fa36322c86da684a211522214bcR33) I think.

File: src/ServiceControl/Connection/PlatformConnectionDetails.cs
Comment (tmasternak) [inline src/ServiceControl/Connection/PlatformConnectionDetails.cs:17]:
```suggestion
            // Add a numeric suffix to duplicated keys e.g. can happen when there are two audit instances configured
```

File: src/ServiceControl/Connection/PlatformConnectionDetails.cs
Comment (tmasternak) [inline src/ServiceControl/Connection/PlatformConnectionDetails.cs:22]: Should we do this only when the key values a different e.g. if there are two audit instances but have the same audit queue name we should not require any user intervention.

File: src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.PublicClr.approved.txt
Comment (tmasternak) [inline src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.PublicClr.approved.txt:270]: This is not used anymore I think.

File: src/ServiceControl/Connection/ConnectionController.cs
Comment (tmasternak) [inline src/ServiceControl/Connection/ConnectionController.cs:20]: Should we have a dedicated type for this response?

File: src/ServiceControl/Connection/RemotePlatformConnectionDetailsProvider.cs
Comment (tmasternak) [inline src/ServiceControl/Connection/RemotePlatformConnectionDetailsProvider.cs:40]:
```suggestion
                        Log.Warn($"Unexpected response from {remote.ApiUri}/connection: ${result}");
```

File: src/ServiceControl/Connection/RemotePlatformConnectionDetailsProvider.cs
Comment (tmasternak) [inline src/ServiceControl/Connection/RemotePlatformConnectionDetailsProvider.cs:51]:
```suggestion
                    var message = $"Unable to get connection details from ServiceControl Audit instance at {remote.ApiUri}/connection.";
```

File: src/ServiceControl/HostingComponent.cs
Comment (tmasternak) [inline src/ServiceControl/HostingComponent.cs:15]: Why transient?

File: src/ServiceControl/Monitoring/HeartbeatsPlatformConnectionDetailsProvider.cs
Comment (tmasternak) [inline src/ServiceControl/Monitoring/HeartbeatsPlatformConnectionDetailsProvider.cs:26]: I think we decided that we should keep this. Right?

**Review by tmasternak — CHANGES_REQUESTED** [review]

File: src/ServiceControl.Audit/Connection/ConnectionController.cs
Comment (mikeminutillo) [reply src/ServiceControl.Audit/Connection/ConnectionController.cs:17]: I added concrete types. There is an approval test for each transport in the multi-instance tests

File: src/ServiceControl/Connection/PlatformConnectionDetails.cs
Comment (mikeminutillo) [reply src/ServiceControl/Connection/PlatformConnectionDetails.cs:22]: The ones which are likely to be duplicates (Message and Saga Audit) are deserialized as a dynamic json object which makes it challenging to determine equality. I think this is an optimization that can come later.

If they aren't equal, then we wouldn't be able to deduplicate, we'll show both, and the customer will have to hand-edit the file to get their desired results.

If they are equal, they we could in theory deduplicate but if we don't then the endpoint will still get details that work.

File: src/ServiceControl.MultiInstance.AcceptanceTests/Connection/PlatformConnectionTests.cs
Comment (mikeminutillo) [reply src/ServiceControl.MultiInstance.AcceptanceTests/Connection/PlatformConnectionTests.cs:88]: Moved the scrubber to the transport integration classes so each one can provide only the scrubbing needed for it's own idiosyncracies

File: src/ServiceControl/Connection/PlatformConnectionDetails.cs
Comment (tmasternak) [reply src/ServiceControl/Connection/PlatformConnectionDetails.cs:22]: Makes sense.

**Review by tmasternak — APPROVED** [review]
image

**Pull request merged**
